homeassistant/components/device_tracker/proximity.py

```diff
@@ -0,0 +1,279 @@
+            else:
+                device_is_closest_to_zone = False
+                _LOGGER.debug('%s: same distance as %s: %s compared with %s', entity_name, device, distance_from_zone, compare_distance_from_zone)
+
+        """========================================================"""
+        #if the device is not the closest to the proximity zone
+        if device_is_closest_to_zone == False:
+            _LOGGER.info('%s: device is not closest to zone', entity_name)
+            return
+
+        """========================================================"""
+        #calculate direction of travel
+        #stop if we cannot calculate the direction of travel (i.e. we don't have a previous state and a current LAT and LONG)
+        if old_state == None or new_state.attributes['latitude'] == None or not('latitude' in old_state.attributes):
+            entity_attributes = {ATTR_DIST_FROM:distance_from_zone, ATTR_DIR_OF_TRAVEL:'Unknown', ATTR_NEAREST_DEVICE:entity_name, ATTR_HIDDEN: False} 
+            hass.states.set(ENTITY_ID, distance_from_zone, entity_attributes)
+            _LOGGER.debug('%s Update entity: distance = %s: direction = unknown: device = %s', ENTITY_ID, distance_from_zone, entity_name)
+            _LOGGER.info('%s: Cannot determine direction of travel as old and/or new LAT or LONG are missing', entity_name)
+            return
+
+        old_distance = distance(proximity_latitude, proximity_longitude, old_state.attributes['latitude'], old_state.attributes['longitude'])
+        _LOGGER.debug('%s: old distance: %s', entity_name, old_distance)
+        new_distance = distance(proximity_latitude, proximity_longitude, new_state.attributes['latitude'], new_state.attributes['longitude'])
+        _LOGGER.debug('%s: new distance: %s', entity_name, new_distance)
+
+        distance_travelled = round(new_distance - old_distance,1)
+
+        #check for a margin of error
+        if distance_travelled <= tolerance * -1:
+            direction_of_travel = 'towards'
+            _LOGGER.info('%s: device travelled %s metres: moving %s', entity_name, distance_travelled, direction_of_travel)
+        elif distance_travelled > tolerance:
+            direction_of_travel = 'away_from'
+            _LOGGER.info('%s: device travelled %s metres: moving %s', entity_name, distance_travelled, direction_of_travel)
+        else:
+            direction_of_travel = 'unknown'
+            _LOGGER.info('%s: Cannot determine direction: %s is too small', entity_name, distance_travelled)
+
+        """========================================================"""
+        #update the proximity entity
+        entity_attributes = {ATTR_DIST_FROM:round(distance_from_zone), ATTR_DIR_OF_TRAVEL:direction_of_travel, ATTR_NEAREST_DEVICE:entity_name, ATTR_HIDDEN: False} 
+        hass.states.set(ENTITY_ID, round(distance_from_zone), entity_attributes)
```

Comment:
You shouldn't use `hass.states.set`. That will be used internally by your Proximity Entity. You would want to update the attributes of your entity and call `proximity.update_ha_state()`:

```python
proximity._dist_from = distance_from_zone
…
proximity.update_ha_state()
```